[FX] Fix NamedTuple literal value printing

jamesr66a opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #49513 [FX] NamedTuple repr approach 3: NTReprWrapper in create_arg
* #49512 [FX] NamedTuple repr approach 2: FakeNamedTuple type in create_arg
* **#49414 [FX] Fix NamedTuple literal value printing**

Previously, when a `NamedTuple` value was used in `args`/`kwargs` and the `NamedTuple` definition was in a different module (necessitating an `import` statement), we would miss 1) Adding the import statement and 2) having the `repr()` that refers to the right fully-qualified name, since the default `repr` for `NamedTuple` in Python does not print the FQN. This patch adds a preprocessing function that swaps out `NamedTuple` instances with an equivalent wrapper that implements `__repr__` to print the right thing.

Differential Revision: [D25564075](https://our.internmc.facebook.com/intern/diff/D25564075)

Commit:
[FX] Fix NamedTuple literal value printing

[ghstack-poisoned]

Commit:
Update on "[FX] Fix NamedTuple literal value printing"


Previously, when a `NamedTuple` value was used in `args`/`kwargs` and the `NamedTuple` definition was in a different module (necessitating an `import` statement), we would miss 1) Adding the import statement and 2) having the `repr()` that refers to the right fully-qualified name, since the default `repr` for `NamedTuple` in Python does not print the FQN. This patch adds a preprocessing function that swaps out `NamedTuple` instances with an equivalent wrapper that implements `__repr__` to print the right thing.

Differential Revision: [D25564075](https://our.internmc.facebook.com/intern/diff/D25564075)

[ghstack-poisoned]

Commit:
Update on "[FX] Fix NamedTuple literal value printing"


Previously, when a `NamedTuple` value was used in `args`/`kwargs` and the `NamedTuple` definition was in a different module (necessitating an `import` statement), we would miss 1) Adding the import statement and 2) having the `repr()` that refers to the right fully-qualified name, since the default `repr` for `NamedTuple` in Python does not print the FQN. This patch adds a preprocessing function that swaps out `NamedTuple` instances with an equivalent wrapper that implements `__repr__` to print the right thing.

Differential Revision: [D25564075](https://our.internmc.facebook.com/intern/diff/D25564075)

[ghstack-poisoned]

Commit:
Update on "[FX] Fix NamedTuple literal value printing"


Previously, when a `NamedTuple` value was used in `args`/`kwargs` and the `NamedTuple` definition was in a different module (necessitating an `import` statement), we would miss 1) Adding the import statement and 2) having the `repr()` that refers to the right fully-qualified name, since the default `repr` for `NamedTuple` in Python does not print the FQN. This patch adds a preprocessing function that swaps out `NamedTuple` instances with an equivalent wrapper that implements `__repr__` to print the right thing.

Differential Revision: [D25564075](https://our.internmc.facebook.com/intern/diff/D25564075)

[ghstack-poisoned]

Commit:
Update on "[FX] Fix NamedTuple literal value printing"


Previously, when a `NamedTuple` value was used in `args`/`kwargs` and the `NamedTuple` definition was in a different module (necessitating an `import` statement), we would miss 1) Adding the import statement and 2) having the `repr()` that refers to the right fully-qualified name, since the default `repr` for `NamedTuple` in Python does not print the FQN. This patch adds a preprocessing function that swaps out `NamedTuple` instances with an equivalent wrapper that implements `__repr__` to print the right thing.

Differential Revision: [D25564075](https://our.internmc.facebook.com/intern/diff/D25564075)

[ghstack-poisoned]

Commit:
Update on "[FX] Fix NamedTuple literal value printing"


Previously, when a `NamedTuple` value was used in `args`/`kwargs` and the `NamedTuple` definition was in a different module (necessitating an `import` statement), we would miss 1) Adding the import statement and 2) having the `repr()` that refers to the right fully-qualified name, since the default `repr` for `NamedTuple` in Python does not print the FQN. This patch adds a preprocessing function that swaps out `NamedTuple` instances with an equivalent wrapper that implements `__repr__` to print the right thing.

Differential Revision: [D25564075](https://our.internmc.facebook.com/intern/diff/D25564075)

[ghstack-poisoned]

Commit:
Update on "[FX] Fix NamedTuple literal value printing"


Previously, when a `NamedTuple` value was used in `args`/`kwargs` and the `NamedTuple` definition was in a different module (necessitating an `import` statement), we would miss 1) Adding the import statement and 2) having the `repr()` that refers to the right fully-qualified name, since the default `repr` for `NamedTuple` in Python does not print the FQN. This patch adds a preprocessing function that swaps out `NamedTuple` instances with an equivalent wrapper that implements `__repr__` to print the right thing.

Differential Revision: [D25564075](https://our.internmc.facebook.com/intern/diff/D25564075)

[ghstack-poisoned]

zdevito commented on torch/fx/graph.py:
```diff
@@ -627,11 +628,52 @@ def delete_unused_values(user : Node):
             else:
                 body.append('\n')
 
+        def preprocess_values(a : Argument):
```
this is a massive copy/paste of map_arg, doesn't seem sustainable.

jamesr66a replied:
This is different from `map_arg` in that the configurable behavior applies to data types other than `Node`. Not sure how to generalize `map_arg` to accept that except getting into weird generic type territory

zdevito commented on torch/fx/graph.py:
```diff
@@ -627,11 +628,52 @@ def delete_unused_values(user : Node):
+            pretty printing
+            """
+            if isinstance(a, tuple) and hasattr(a, '_fields'):
+                # Annoyingly, Python NamedTuples repr() with an unqualified
```
I don't know if supporting named tuples in arguments is worth this complexity. Can we just capture something simpler than the named tuple and convert named tuples to it? I am pretty sure this isn't the first patch fixing named tuples and seems to be adding more complexity than the value it adds.

```
class Tuple:
   def __init__(self, **kwargs):
     self.__dict__.update(kwargs)
   def __repr__(self):
       return f"torch.fx.Tuple(...)"
   def __iter__(self):
        return iter(self.values())
   def __getitem__(self, i):
        return self.values()[i]
```

jamesr66a replied:
This patch is primarily aimed at the case where a `NamedTuple` is the return from a given `Module`, which is an extremely common pattern. I don't think returning a generic tuple is going to work, since surrounding code that calls the module may expect the object returned to be indexable by name

Commit:
Update on "[FX] Fix NamedTuple literal value printing"


Previously, when a `NamedTuple` value was used in `args`/`kwargs` and the `NamedTuple` definition was in a different module (necessitating an `import` statement), we would miss 1) Adding the import statement and 2) having the `repr()` that refers to the right fully-qualified name, since the default `repr` for `NamedTuple` in Python does not print the FQN. This patch adds a preprocessing function that swaps out `NamedTuple` instances with an equivalent wrapper that implements `__repr__` to print the right thing.

Differential Revision: [D25564075](https://our.internmc.facebook.com/intern/diff/D25564075)

[ghstack-poisoned]
